### PR TITLE
GPII-2959: Adding nginx-ingress, cert-manager charts, updating flowmanager and preferences services, fixes for new deploy order

### PR DESCRIPTION
**Pull request by natarajaya**

### CODE REVIEW

File: aws/rakefiles/deploy.rake [hunk 1]
```diff
@@ -107,18 +106,19 @@ task :deploy_only => [:configure_kubectl, :install_charts, :find_gpii_components
...
 
   installed_charts = `helm list -q -a`
   installed_charts = installed_charts.split("\n")
   @gpii_helmcharts.each do |chart|
+    chart_config = YAML.load_file("#{@tmpdir}-modules/deploy/helms/#{chart}/custom-values.yaml")
     if installed_charts.include?(chart)
       begin
         wait_for(
-          "helm upgrade --recreate-pods -f #{@tmpdir}-modules/deploy/helms/#{chart}/custom-values.yaml #{chart} #{@tmpdir}-modules/deploy/helms/#{chart}",
+          "helm upgrade --recreate-pods -f #{@tmpdir}-modules/deploy/helms/#{chart}/custom-values.yaml #{chart_config['helm-override'] && chart_config['helm-override']['name'] ? chart_config['helm-override']['name'] : chart} #{@tmpdir}-modules/deploy/helms/#{chart}",
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply aws/rakefiles/deploy.rake]: Done

File: aws/rakefiles/deploy.rake [hunk 2]
```diff
@@ -128,7 +128,7 @@ task :install_charts => [:configure_kubectl, :setup_system_components, :init_hel
     else
       begin
         wait_for(
-          "helm install --name #{chart} -f #{@tmpdir}-modules/deploy/helms/#{chart}/custom-values.yaml #{@tmpdir}-modules/deploy/helms/#{chart}",
+          "helm install --name #{chart_config['helm-override'] && chart_config['helm-override']['name'] ? chart_config['helm-override']['name'] : chart} #{chart_config['helm-override'] && chart_config['helm-override']['namespace'] ? "--namespace " + chart_config['helm-override']['namespace'] : ""} -f #{@tmpdir}-modules/deploy/helms/#{chart}/custom-values.yaml #{@tmpdir}-modules/deploy/helms/#{chart}",
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply aws/rakefiles/deploy.rake]: Done

File: aws/modules/deploy/helms/nginx-ingress-values.erb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+helm-override:
+  name: nginx-ingress
+  namespace: gpii
+
+rbac:
```
[the remark this answers is not in the document]
Comment (mrtyler) [reply aws/modules/deploy/helms/nginx-ingress-values.erb]: Sorry, I still don't understand this.

File: aws/modules/deploy/Rakefile [hunk 4]
```diff
@@ -130,7 +130,7 @@ do
   end
 
   HELMCHARTS_DIRS.each do |d|
-    rendered = ERB.new(File.read("helms/#{d}-values.erb")).result(binding)
+    rendered = ERB.new(File.read("helms/#{d}-values.#{File.file?("helms/#{d}-values.erb") ? 'erb' : 'yml'}")).result(binding)
```
Comment (mrtyler) [inline aws/modules/deploy/Rakefile]: Hmm, I didn't understand how this was set up. Let's get rid of this extra logic and rename the file back to .erb. Sorry for the confusion.

(It would be better if this used the logic below for `rule ".yml" => ".erb"` but let's not mess with that today.)

File: aws/modules/deploy/helms/cert-manager-values.yml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+chart-metadata:
```
Comment (mrtyler) [inline aws/modules/deploy/helms/cert-manager-values.yml]: This reads better to me, thanks.

I still don't like the name of this file. How about cert-manager-metadata.yml or cert-manager-gpii-infra.yml or something to indicate that this is a special, non-Helm facility.

File: aws/rakefiles/deploy.rake [hunk 6]
```diff
@@ -17,20 +17,14 @@ desc "Wait until GPII components have been deployed"
...
-  # It's not perfect but it's a good start.
   preferences_url = "https://preferences.#{ENV["TF_VAR_cluster_name"]}/preferences/carla"
-  if ENV["TF_VAR_cluster_name"].start_with?("stg.", "prd.")
-    preferences_url.gsub! "http://", "https://"
+  if ENV["TF_VAR_cluster_name"].start_with?("prd.")
+    wait_for("curl --silent --output /dev/stderr --write-out '%{http_code}' '#{preferences_url}' | grep -q ^2")
+  else
+    wait_for("curl -k --silent --output /dev/stderr --write-out '%{http_code}' '#{preferences_url}' | grep -q ^2")
+    # For staging and dev environments we also need to make sure that certificate is issues by Letsencrypt
+    wait_for("curl -k -vI https://preferences.dev-natarajaya.gpii.net/preferences/carla 2>&1 | grep 'CN=Fake LE Intermediate X1'")
```
Comment (mrtyler) [inline aws/rakefiles/deploy.rake]: Hardcoded dev-natarajaya should be TF_VAR_cluster_name.

Comment (mrtyler) [inline aws/rakefiles/deploy.rake]: By default, wait_for will try this command every 20 seconds for 15 minutes. Are those defaults reasonable for this command, especially since if we get to this line then we know the previous command succeeded?

File: aws/rakefiles/deploy.rake [hunk 7]
```diff
@@ -107,18 +106,19 @@ task :deploy_only => [:configure_kubectl, :install_charts, :find_gpii_components
...
-task :install_charts => [:configure_kubectl, :setup_system_components, :init_helm] do
+task :install_charts => [:configure_kubectl, :generate_modules, :setup_system_components, :init_helm] do
   Dir.chdir("#{@tmpdir}-modules/deploy/helms/") do
     @gpii_helmcharts = Dir.glob("*").select {|d| File.directory? d }
   end
 
   installed_charts = `helm list -q -a`
   installed_charts = installed_charts.split("\n")
   @gpii_helmcharts.each do |chart|
+    chart_config = YAML.load_file("#{@tmpdir}-modules/deploy/helms/#{chart}/custom-values.yaml")
```
[the remark this answers is not in the document]
Comment (mrtyler) [reply aws/rakefiles/deploy.rake]: Ah, ok.

File: aws/rakefiles/deploy.rake [hunk 8]
```diff
@@ -115,10 +109,21 @@ task :install_charts => [:configure_kubectl, :generate_modules, :setup_system_co
...
+    if chart_config['chart-metadata'] && chart_config['chart-metadata']['namespace']
+      chart_namespace = chart_config['chart-metadata']['namespace']
+    else
+      chart_namespace = 'default'
+    end
     if installed_charts.include?(chart)
       begin
         wait_for(
-          "helm upgrade --recreate-pods -f #{@tmpdir}-modules/deploy/helms/#{chart}/custom-values.yaml #{chart_config['helm-override'] && chart_config['helm-override']['name'] ? chart_config['helm-override']['name'] : chart} #{@tmpdir}-modules/deploy/helms/#{chart}",
+          "helm upgrade --recreate-pods -f #{@tmpdir}-modules/deploy/helms/#{chart}/custom-values.yaml #{chart_name} #{@tmpdir}-modules/deploy/helms/#{chart}",
```
Comment (mrtyler) [inline aws/rakefiles/deploy.rake]: Can we include --namespace here for consistency with the `install` path?

File: aws/rakefiles/deploy.rake (hunk 6 above)
Comment (mrtyler) [reply aws/rakefiles/deploy.rake]: Ah, you fixed this in the next commit. Never mind :)

**Comment by mrtyler** [comment]
Github is confused about some things:
* Some of your comments showed up as responses to my initial review (this is what I consider "normal" and desirable) but others showed up in a separate review. I'm not sure what happened but it's a little hard to follow some of the discussion.
* Because I often like to go commit-by-commit, I added some review comments on 0bf6097af56ee1bec76a1bc63aa2cb7ce7c8f842 which are showing as "outdated" because of the subsequent commit 156ba35e8b88ae83369a5d0bff9f5bac04694bde. So please make sure to view all my comments even if Github thinks they're "outdated".

File: aws/rakefiles/deploy.rake [hunk 9]
```diff
@@ -17,20 +17,14 @@ desc "Wait until GPII components have been deployed"
 task :wait_for_gpii_ready => :configure_kubectl do
   puts "Waiting for GPII components to be fully deployed..."
   puts "(You can Ctrl-C out of this safely. You may need to re-run :deploy_only afterward.)"
-
-  # This is the simplest one-liner I could find to GET a url and return just
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply aws/rakefiles/deploy.rake]: Ok, lets put it back!

File: aws/rakefiles/deploy.rake (hunk 6 above)
Comment (natarajaya) [reply aws/rakefiles/deploy.rake]: Changed to:
```    wait_for(
      "curl -k -vI #{preferences_url} 2>&1 | grep 'CN=Fake LE Intermediate X1'",
      sleep_secs: 5,
      max_wait_secs: 20,
    )```

File: aws/modules/deploy/Rakefile (hunk 4 above)
Comment (natarajaya) [reply aws/modules/deploy/Rakefile]: Reverted

File: aws/rakefiles/deploy.rake (hunk 6 above)
Comment (natarajaya) [reply aws/rakefiles/deploy.rake]: Makes sense

File: aws/rakefiles/deploy.rake (hunk 8 above)
Comment (natarajaya) [reply aws/rakefiles/deploy.rake]: Done

**Comment by mrtyler** [comment]
Sergey and I chatted and he explained what's going on with `\*-values.yaml` files.

I have one last request, of course :): Can you please rename `{chart}-values.yaml` to `{chart}-custom-values.yaml`? This better matches the name of the generated file (`custom-values.yaml`), and the word "custom" is a small clue that this is not a vanilla Helm mechanism.

After that, this LGTM!

File: aws/modules/deploy/helms/README.md [hunk 10]
```diff
@@ -5,9 +5,14 @@ This is the place where the GPII Helm Charts live. The structure is pretty
 basic:
 
 - A directory per Helm Chart
-- A variable file that will override the default variables of the chart with the naming (Helm Char Name)-values.erb
+- A variable file that will override the default variables of the chart with the naming (Helm Char Name)-custom-values.erb
```
Comment (mrtyler) [inline aws/modules/deploy/helms/README.md]: Helm Chart* Name :)

Comment (natarajaya) [reply aws/modules/deploy/helms/README.md]: Doh

**Comment by mrtyler** [comment]
Nice, thanks for adding that note to the README.

LGTM!
